pkg/inventory/inventory-client.go

```diff
@@ -6,473 +6,402 @@ package inventory
-	if err != nil {
-		return objs, err
-	}
-	return wrapped.Load()
-}
-
-// getClusterInventoryObj returns a pointer to the cluster inventory object, or
-// an error if one occurred. Returns the cached cluster inventory object if it
-// has been previously retrieved. Uses the ResourceBuilder to retrieve the
-// inventory object in the cluster, using the namespace, group resource, and
-// inventory label. Merges multiple inventory objects into one if it retrieves
-// more than one (this should be very rare).
-//
-// TODO(seans3): Remove the special case code to merge multiple cluster inventory
-// objects once we've determined that this case is no longer possible.
-func (cic *ClusterClient) getClusterInventoryInfo(ctx context.Context, inv Info) (*unstructured.Unstructured, error) {
-	clusterInvObjects, err := cic.getClusterInventoryObjs(ctx, inv)
-	if err != nil {
-		return nil, fmt.Errorf("failed to read inventory objects from cluster: %w", err)
-	}
+type WriteClient interface {
+	CreateOrUpdate(ctx context.Context, inv Inventory, opts UpdateOptions) error
+	Delete(ctx context.Context, inv Info, opts DeleteOptions) error
+}
 
-	var clusterInv *unstructured.Unstructured
-	if len(clusterInvObjects) == 1 {
-		clusterInv = clusterInvObjects[0]
-	} else if l := len(clusterInvObjects); l > 1 {
-		return nil, fmt.Errorf("found %d inventory objects with inventory id %s", l, inv.ID())
-	}
-	return clusterInv, nil
+type UpdateOptions struct{}
+
+type GetOptions struct{}
+
+type ListOptions struct{}
+
+type DeleteOptions struct{}
+
+var _ Client = &UnstructuredClient{}
+
+// UnstructuredInventory implements Inventory while also tracking the actual
+// KRM object from the cluster. This enables the client to update the
+// same object and utilize resourceVersion checks.
+type UnstructuredInventory struct {
+	BaseInventory
+	// ClusterObj represents the KRM which was last fetched from the cluster.
+	// used by the client implementation to performs updates on the object.
+	ClusterObj *unstructured.Unstructured
 }
 
-func (cic *ClusterClient) getClusterInventoryObjsByLabel(ctx context.Context, inv Info) (object.UnstructuredSet, error) {
-	localInv, err := cic.invToUnstructuredFunc(inv)
-	if err != nil {
-		return nil, err
+func (ui *UnstructuredInventory) Info() Info {
+	if ui.ClusterObj == nil {
+		return SingleObjectInfo{}
 	}
-	if localInv == nil {
-		return nil, fmt.Errorf("retrieving cluster inventory object with nil local inventory")
-	}
-	localObj := object.UnstructuredToObjMetadata(localInv)
-	mapping, err := cic.getMapping(localInv)
-	if err != nil {
-		return nil, err
-	}
-	groupResource := mapping.Resource.GroupResource().String()
-	namespace := localObj.Namespace
-	label, err := retrieveInventoryLabel(localInv)
-	if err != nil {
-		return nil, err
+	return NewSingleObjectInfo(ui.ID(), client.ObjectKeyFromObject(ui.ClusterObj),
+		maps.Clone(ui.ClusterObj.GetLabels()), maps.Clone(ui.ClusterObj.GetAnnotations()))
```

Comment:
I'm torn between a few variants of this:
- use maps.Clone here defensively, but expensively
- pre-compute the Info at construction time & return a DeepCopy here
- pre-compute the Info at construction time & return the pre-computed Info here, requiring the caller to DeepCopy if they want to pass it somewhere else
- lazily compute the Info and cache it, then return a DeepCopy here
- lazily compute the Info and cache it, then return the lazily-computed Info here

Reply:
Nevermind. We solve this by removing labels and annotations from the Info.